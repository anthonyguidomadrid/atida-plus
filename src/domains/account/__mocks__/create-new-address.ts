export const createNewAddress = {
  redirect: 'some-path',
  salutation: 'string',
  firstName: 'string',
  lastName: 'string',
  address1: 'string',
  address3: 'string',
  zipCode: 'string',
  city: 'string',
  country: 'string',
  iso2Code: 'string',
  phone: 'string',
  isDefaultShipping: true,
  isDefaultBilling: true
}

export const createNewAddressPayload = {
  data: {
    type: 'customers',
    attributes: {
      salutation: 'string',
      firstName: 'string',
      lastName: 'string',
      address1: 'string',
      address3: 'string',
      zipCode: 'string',
      city: 'string',
      country: 'string',
      iso2Code: 'string',
      phone: 'string',
      isDefaultShipping: true,
      isDefaultBilling: true
    }
  }
} as const

export const UpdatedJWTAndRefreshTokens = {
  refreshToken: 'newRefreshToken',
  JWT: 'newJWT',
  expiresIn: '12345'
}

export const sprykerUpdatedJWTAndRefreshTokens = {
  data: {
    type: '',
    id: '',
    attributes: {
      tokenType: 'Bearer',
      expiresIn: '12345',
      accessToken: 'newJWT',
      refreshToken: 'newRefreshToken',
      idCompanyUser: ''
    },
    links: {
      self: ''
    }
  },
  links: {
    self: ''
  }
}
