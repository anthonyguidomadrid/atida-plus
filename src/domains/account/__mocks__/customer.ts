export const createCustomer = {
  salutation: 'Mr',
  firstName: 'Test',
  lastName: 'McTest',
  email: 'some@email.com',
  password: 'somePassword1!',
  confirmPassword: 'somePassword1!',
  acceptedTerms: true,
  emailNotification: true,
  phone: '654654654',
  company: null,
  surcharge: false,
  addresses: {
    addresses: [
      {
        salutation: 'Mr',
        firstName: 'Test',
        lastName: 'McTest',
        address1: 'Brugstraat 123',
        zipCode: '1011 AD',
        city: 'Amsterdam',
        iso2Code: 'PT',
        phone: '088273791298432',
        isDefaultShipping: true,
        isDefaultBilling: true
      }
    ]
  }
}

export const sprykerCustomer = {
  data: {
    type: 'customers',
    id: 'some-id',
    attributes: {
      salutation: 'Mr',
      firstName: 'Test',
      lastName: 'McTest',
      email: 'some@email.com',
      password: 'somePassword1!',
      confirmPassword: 'somePassword1!',
      acceptedTerms: true,
      emailNotification: true,
      company: null,
      surcharge: false
    }
  }
} as const

export const sprykerCustomerWithAddresses = {
  data: {
    ...sprykerCustomer,
    attributes: {
      ...sprykerCustomer.data.attributes
    },
    relationships: {
      addresses: {
        data: [
          { type: 'addresses', id: 'address-id-1' },
          { type: 'addresses', id: 'address-id-2' }
        ]
      }
    }
  },
  included: [
    {
      type: 'addresses',
      id: 'address-id-1',
      attributes: {
        salutation: 'some-salutation',
        firstName: 'some-name',
        lastName: 'some-last-name',
        address1: 'some-address',
        address2: 'some-address2',
        address3: 'some-address3',
        zipCode: 'some-zip-code',
        city: 'some-city',
        country: 'some-country',
        iso2Code: 'some-iso-code',
        company: 'some-company',
        phone: 'some-phone',
        isDefaultShipping: true,
        isDefaultBilling: true,
        isTaxExempt: false
      }
    },
    {
      type: 'addresses',
      id: 'address-id-2',
      attributes: {
        salutation: 'some-other-salutation',
        firstName: 'some-other-name',
        lastName: 'some-other-last-name',
        address1: 'some-other-address',
        address2: 'some-other-address2',
        address3: 'some-other-address3',
        zipCode: 'some-other-zip-code',
        city: 'some-other-city',
        country: 'some-other-country',
        iso2Code: 'some-other-iso-code',
        company: 'some-other-company',
        phone: 'some-other-phone',
        isDefaultShipping: false,
        isDefaultBilling: false,
        isTaxExempt: false
      }
    }
  ]
}
