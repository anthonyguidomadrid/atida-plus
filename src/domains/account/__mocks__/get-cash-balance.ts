export const balance = {
  total: 20
}
