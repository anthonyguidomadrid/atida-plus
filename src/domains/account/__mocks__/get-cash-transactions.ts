import { AtidaCashTransactionHistoryLedgerResponse } from '../types'

export const mockedTransactionHistory = {
  data: [
    {
      transaction_type: 'debit',
      amount: 200,
      transaction_date: '2022-10-04T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 280,
      expiry_date: '2023-10-04T14:20:22.029764445Z',
      transaction_date: '2022-10-04T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 280,
      expiry_date: '2023-07-23T14:20:22.029764445Z',
      transaction_date: '2022-07-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'debit',
      amount: 200,
      transaction_date: '2022-06-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 280,
      expiry_date: '2023-06-23T14:20:22.029764445Z',
      transaction_date: '2022-06-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 500,
      expiry_date: '2023-05-23T14:20:22.029764445Z',
      transaction_date: '2022-05-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'debit',
      amount: 280,
      expiry_date: '2023-04-23T14:20:22.029764445Z',
      transaction_date: '2022-04-23T14:20:22.029764445Z'
    },
    {
      transaction_type: 'credit',
      amount: 200,
      expiry_date: '2023-04-23T14:20:22.029764445Z',
      transaction_date: '2022-04-23T14:20:22.029764445Z'
    }
  ],
  metadata: {
    count: 8,
    total: 27
  }
} as AtidaCashTransactionHistoryLedgerResponse

export const mockedTransactionHistoryBayDateArray = {
  transactions: [
    {
      date: '2022-09-30',
      transactions: [
        {
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'debit',
          amount: 200
        },
        {
          amount: 280,
          expiry_date: '2023-09-30T13:36:27.297Z',
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'credit'
        },
        {
          transaction_type: 'debit',
          amount: 200,
          transaction_date: '2022-09-30T13:36:27.297Z'
        },
        {
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'credit',
          amount: 180,
          expiry_date: '2023-09-30T13:36:27.297Z'
        },
        {
          amount: 100,
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'debit'
        },
        {
          transaction_date: '2022-09-30T13:36:27.297Z',
          transaction_type: 'credit',
          amount: 670,
          expiry_date: '2023-09-30T13:36:27.297Z'
        }
      ]
    },
    {
      date: '2022-07-23',
      transactions: [
        {
          transaction_date: '2022-07-23T14:20:22.029764445Z',
          transaction_type: 'debit',
          amount: 200
        },
        {
          expiry_date: '2023-07-23T14:20:22.029764445Z',
          transaction_date: '2022-07-23T14:20:22.029764445Z',
          transaction_type: 'credit',
          amount: 280
        },
        {
          transaction_date: '2022-07-23T14:20:22.029764445Z',
          transaction_type: 'debit',
          amount: 200
        },
        {
          amount: 290,
          expiry_date: '2023-07-23T14:20:22.029764445Z',
          transaction_date: '2022-07-23T14:20:22.029764445Z',
          transaction_type: 'credit'
        }
      ]
    },
    {
      date: '2022-06-23',
      transactions: [
        {
          transaction_type: 'debit',
          amount: 200,
          transaction_date: '2022-06-23T14:20:22.029764445Z'
        },
        {
          expiry_date: '2023-06-23T14:20:22.029764445Z',
          transaction_date: '2022-06-23T14:20:22.029764445Z',
          transaction_type: 'credit',
          amount: 280
        }
      ]
    },
    {
      date: '2022-05-23',
      transactions: [
        {
          transaction_type: 'credit',
          amount: 500,
          expiry_date: '2023-05-23T14:20:22.029764445Z',
          transaction_date: '2022-05-23T14:20:22.029764445Z'
        }
      ]
    },
    {
      date: '2022-04-23',
      transactions: [
        {
          transaction_date: '2022-04-23T14:20:22.029764445Z',
          transaction_type: 'debit',
          amount: 280,
          expiry_date: '2023-04-23T14:20:22.029764445Z'
        },
        {
          transaction_type: 'credit',
          amount: 200,
          expiry_date: '2023-04-23T14:20:22.029764445Z',
          transaction_date: '2022-04-23T14:20:22.029764445Z'
        }
      ]
    }
  ],
  totalPages: 4
}
