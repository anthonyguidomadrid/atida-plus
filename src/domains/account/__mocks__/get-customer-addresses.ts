import { SprykerCustomerAddresses, CustomerAddress } from '../types'

export const sprykerCustomerAddresses: SprykerCustomerAddresses = {
  data: [
    {
      type: 'addresses',
      id: 'some-id',
      attributes: {
        salutation: 'some-salutation',
        firstName: 'some-name',
        lastName: 'some-last-name',
        address1: 'some-address',
        address2: 'some-address2',
        address3: 'some-address3',
        zipCode: 'some-zip-code',
        city: 'some-city',
        country: 'some-country',
        iso2Code: 'some-iso-code',
        company: 'some-company',
        phone: 'some-phone',
        createdAt: '2021-12-27 14:52:50.392774',
        updatedAt: '2021-12-27 14:52:50.392774',
        isDefaultShipping: true,
        isDefaultBilling: true,
        isTaxExempt: false
      }
    },
    {
      type: 'addresses',
      id: 'some-other-id',
      attributes: {
        salutation: 'some-other-salutation',
        firstName: 'some-other-name',
        lastName: 'some-other-last-name',
        address1: 'some-other-address',
        address2: 'some-other-address2',
        address3: 'some-other-address3',
        zipCode: 'some-other-zip-code',
        city: 'some-other-city',
        country: 'some-other-country',
        iso2Code: 'some-other-iso-code',
        company: 'some-other-company',
        phone: 'some-other-phone',
        createdAt: '2021-12-27 14:52:50.392774',
        updatedAt: '2021-12-27 14:52:50.392774',
        isDefaultShipping: false,
        isDefaultBilling: false,
        isTaxExempt: false
      }
    }
  ]
}

/*
This address list is used to test the AddressList sorting functionality

Priority ordering (with the reason for that priority):
  P0 > P1 (shipping over billing)
  P1 > P2 (billing over availability)
  P2 > P3 (availability over recently updated addresses)
  P3 > P4 (recently updated addresses over 'stale' addresses)
*/
export const unsortedCustomerAddresses: CustomerAddress[] = [
  {
    id: 'some-id',
    salutation: 'some-salutation',
    firstName: '[priority-through-shipping]',
    lastName: 'some-last-name',
    address1: 'P0',
    address2: 'some-address2',
    address3: 'some-address3',
    zipCode: 'some-zip-code',
    city: 'some-city',
    country: 'Spain',
    iso2Code: 'ES',
    company: 'some-company',
    phone: 'some-phone',
    createdAt: '2021-12-27 14:52:50.392774',
    updatedAt: '2021-12-27 14:52:50.392774',
    isDefaultShipping: true,
    isDefaultBilling: false,
    isTaxExempt: false
  },
  {
    id: 'some-other-id',
    salutation: 'some-other-salutation',
    firstName: '[priority-through-billing]',
    lastName: 'some-other-last-name',
    address1: 'P1',
    address2: 'some-other-address2',
    address3: 'some-other-address3',
    zipCode: 'some-other-zip-code',
    city: 'some-other-city',
    country: 'some-other-country',
    iso2Code: 'ES',
    company: 'Spain',
    phone: 'some-other-phone',
    createdAt: '2021-12-27 14:52:50.392774',
    updatedAt: '2021-12-27 14:52:50.392774',
    isDefaultShipping: false,
    isDefaultBilling: true,
    isTaxExempt: false
  },
  {
    id: 'some-other-id',
    salutation: 'some-other-salutation',
    firstName: '[priority-through-availability]',
    lastName: 'some-other-last-name',
    address1: 'P2',
    address2: 'some-other-address2',
    address3: 'some-other-address3',
    zipCode: 'some-other-zip-code',
    city: 'some-other-city',
    country: 'Spain',
    iso2Code: 'ES',
    company: 'some-other-company',
    phone: 'some-other-phone',
    createdAt: '2021-12-27 14:52:50.392774',
    updatedAt: '2021-12-27 14:52:50.392774',
    isDefaultShipping: false,
    isDefaultBilling: false,
    isTaxExempt: false
  },
  {
    id: 'some-other-id',
    salutation: 'some-other-salutation',
    firstName: '[priority-through-recency]',
    lastName: 'some-other-last-name',
    address1: 'P3',
    address2: 'some-other-address2',
    address3: 'some-other-address3',
    zipCode: 'some-other-zip-code',
    city: 'some-other-city',
    country: 'Portugal',
    iso2Code: 'PT',
    company: 'some-other-company',
    phone: 'some-other-phone',
    createdAt: '2021-12-28 14:52:50.392774',
    updatedAt: '2021-12-28 14:52:50.392774',
    isDefaultShipping: false,
    isDefaultBilling: false,
    isTaxExempt: false
  },
  {
    id: 'some-other-id',
    salutation: 'some-other-salutation',
    firstName: '[lowest-priority]',
    lastName: 'some-other-last-name',
    address1: 'P4',
    address2: 'some-other-address2',
    address3: 'some-other-address3',
    zipCode: 'some-other-zip-code',
    city: 'some-other-city',
    country: 'Portugal',
    iso2Code: 'PT',
    company: 'some-other-company',
    phone: 'some-other-phone',
    createdAt: '2021-12-27 14:52:50.392774',
    updatedAt: '2021-12-27 14:52:50.392774',
    isDefaultShipping: false,
    isDefaultBilling: false,
    isTaxExempt: false
  }
]

/*
This address list is used to test the AddressList sorting functionality

Priority ordering (with the reason for that priority):
  P0 > P1 (shipping over billing)
  P1 > P2 (billing over availability)
  P2 > P3 (availability over recently updated addresses)
  P3 > P4 (recently updated addresses over 'stale' addresses)
*/
export const unsortedCustomerAddressesISO801Compatible: CustomerAddress[] = [
  {
    id: 'some-id',
    salutation: 'some-salutation',
    firstName: '[priority-through-shipping]',
    lastName: 'some-last-name',
    address1: 'P0',
    address2: 'some-address2',
    address3: 'some-address3',
    zipCode: 'some-zip-code',
    city: 'some-city',
    country: 'Spain',
    iso2Code: 'ES',
    company: 'some-company',
    phone: 'some-phone',
    createdAt: '2021-12-27T14:52:50.392774',
    updatedAt: '2021-12-27T14:52:50.392774',
    isDefaultShipping: true,
    isDefaultBilling: false,
    isTaxExempt: false
  },
  {
    id: 'some-other-id',
    salutation: 'some-other-salutation',
    firstName: '[priority-through-billing]',
    lastName: 'some-other-last-name',
    address1: 'P1',
    address2: 'some-other-address2',
    address3: 'some-other-address3',
    zipCode: 'some-other-zip-code',
    city: 'some-other-city',
    country: 'some-other-country',
    iso2Code: 'ES',
    company: 'Spain',
    phone: 'some-other-phone',
    createdAt: '2021-12-27T14:52:50.392774',
    updatedAt: '2021-12-27T14:52:50.392774',
    isDefaultShipping: false,
    isDefaultBilling: true,
    isTaxExempt: false
  },
  {
    id: 'some-other-id',
    salutation: 'some-other-salutation',
    firstName: '[priority-through-availability]',
    lastName: 'some-other-last-name',
    address1: 'P2',
    address2: 'some-other-address2',
    address3: 'some-other-address3',
    zipCode: 'some-other-zip-code',
    city: 'some-other-city',
    country: 'Spain',
    iso2Code: 'ES',
    company: 'some-other-company',
    phone: 'some-other-phone',
    createdAt: '2021-12-27T14:52:50.392774',
    updatedAt: '2021-12-27T14:52:50.392774',
    isDefaultShipping: false,
    isDefaultBilling: false,
    isTaxExempt: false
  },
  {
    id: 'some-other-id',
    salutation: 'some-other-salutation',
    firstName: '[priority-through-recency]',
    lastName: 'some-other-last-name',
    address1: 'P3',
    address2: 'some-other-address2',
    address3: 'some-other-address3',
    zipCode: 'some-other-zip-code',
    city: 'some-other-city',
    country: 'Portugal',
    iso2Code: 'PT',
    company: 'some-other-company',
    phone: 'some-other-phone',
    createdAt: '2021-12-28T14:52:50.392774',
    updatedAt: '2021-12-28T14:52:50.392774',
    isDefaultShipping: false,
    isDefaultBilling: false,
    isTaxExempt: false
  },
  {
    id: 'some-other-id',
    salutation: 'some-other-salutation',
    firstName: '[lowest-priority]',
    lastName: 'some-other-last-name',
    address1: 'P4',
    address2: 'some-other-address2',
    address3: 'some-other-address3',
    zipCode: 'some-other-zip-code',
    city: 'some-other-city',
    country: 'Portugal',
    iso2Code: 'PT',
    company: 'some-other-company',
    phone: 'some-other-phone',
    createdAt: '2021-12-27T14:52:50.392774',
    updatedAt: '2021-12-27T14:52:50.392774',
    isDefaultShipping: false,
    isDefaultBilling: false,
    isTaxExempt: false
  }
]

export const Addresses: CustomerAddress[] = [
  {
    id: 'some-id',
    salutation: 'some-salutation',
    firstName: '[priority-through-shipping]',
    lastName: 'some-last-name',
    address1: 'P0',
    address2: 'some-address2',
    address3: 'some-address3',
    zipCode: 'some-zip-code',
    city: 'some-city',
    country: 'Portugal',
    iso2Code: 'PT',
    company: 'some-company',
    phone: 'some-phone',
    createdAt: '2021-12-27 14:52:50.392774',
    updatedAt: '2021-12-27 14:52:50.392774',
    isDefaultShipping: true,
    isDefaultBilling: false,
    isTaxExempt: false
  },
  {
    id: 'some-other-id',
    salutation: 'some-other-salutation',
    firstName: '[priority-through-billing]',
    lastName: 'some-other-last-name',
    address1: 'P1',
    address2: 'some-other-address2',
    address3: 'some-other-address3',
    zipCode: 'some-other-zip-code',
    city: 'some-other-city',
    country: 'some-other-country',
    iso2Code: 'PT',
    company: 'Portugal',
    phone: 'some-other-phone',
    createdAt: '2021-12-27 14:52:50.392774',
    updatedAt: '2021-12-27 14:52:50.392774',
    isDefaultShipping: false,
    isDefaultBilling: true,
    isTaxExempt: false
  }
]
