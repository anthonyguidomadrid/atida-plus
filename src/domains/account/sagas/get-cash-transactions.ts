import {
  AtidaCashTransactionsHistoryByDate,
  AtidaCashTransactionTriggerPayload
} from './../types'
import { AxiosResponse } from 'axios'
import { all, call, getContext, put, takeLatest } from 'redux-saga/effects'
import {
  getCashTransactionsTrigger,
  getCashTransactionsRequest,
  getCashTransactionsSuccess,
  getCashTransactionsFailure,
  getCashTransactionsFulfill
} from '../slices'
import { createClient, logger } from '~helpers'
import { getErrorMessage } from '~helpers/error'
import type { SagaIterator } from 'redux-saga'
import type { PayloadAction } from '@reduxjs/toolkit'

function* doGetCashTransactionsSaga({
  payload
}: PayloadAction<AtidaCashTransactionTriggerPayload>) {
  try {
    yield put(getCashTransactionsRequest())
    const locale: string = yield getContext('locale')
    const client = createClient({ locale })
    const response: AxiosResponse<AtidaCashTransactionsHistoryByDate> = yield call(
      client.get,
      `/api/account/cash-transactions`,
      { params: payload }
    )
    yield put(getCashTransactionsSuccess(response.data))
  } catch (error) {
    logger.error(getErrorMessage(error))
    yield put(
      getCashTransactionsFailure({
        message: getErrorMessage(error)
      })
    )
  } finally {
    yield put(getCashTransactionsFulfill())
  }
}

export function* getCashTransactionsSaga(): SagaIterator {
  yield all([takeLatest(getCashTransactionsTrigger, doGetCashTransactionsSaga)])
}
