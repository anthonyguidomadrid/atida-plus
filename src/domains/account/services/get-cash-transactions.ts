import getConfig from 'next/config'
import { FeatureFlag } from '~config/constants/feature-flags'
import { createClient } from '~helpers'
import { loadFeatureFlag } from '~helpers/server-only/featureFlagClient'
import { normalizeTransactions } from '../normalizers/transactions'
import {
  AtidaCashTransactionHistoryLedgerResponse,
  AtidaCashTransactionRequestPayload,
  AtidaCashTransactionsHistory
} from '../types'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

export const getCashTransactions = async ({
  locale,
  params,
  accessToken
}: AtidaCashTransactionRequestPayload): Promise<AtidaCashTransactionsHistory> => {
  const useMockedData = (await loadFeatureFlag(
    locale,
    FeatureFlag.ACCOUNT_TRANSACTION_HISTORY_MOCKED_DATA
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  )) as any
  if (Object.keys(useMockedData).length > 0) {
    let totalPages = 0
    if (useMockedData.metadata.total && params?.limit) {
      totalPages = Math.ceil(useMockedData.metadata.total / +params.limit)
    }
    return {
      transactions: normalizeTransactions(useMockedData) ?? [],
      totalPages
    }
  }
  try {
    const { serverRuntimeConfig } = getConfig()

    const client = createClient({
      locale,
      options: { baseURL: serverRuntimeConfig.ledgerBaseUrl }
    })

    const response = await client.get<AtidaCashTransactionHistoryLedgerResponse>(
      `/ledger-data?limit=${params?.limit}&offset=${params?.offset}`,
      {
        headers: {
          Authorization: `Bearer ${accessToken}`
        }
      }
    )

    const transactions = response.data
    let totalPages = 0
    if (response.data.metadata.total && params?.limit) {
      totalPages = Math.ceil(response.data.metadata.total / +params.limit)
    }
    return {
      transactions: normalizeTransactions(transactions) ?? [],
      totalPages
    }
  } catch (error) {
    return {
      transactions: [],
      totalPages: 0
    }
  }
}
