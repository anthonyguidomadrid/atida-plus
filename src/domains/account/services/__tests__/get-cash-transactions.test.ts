/**
 * @jest-environment node
 */
import axios from 'axios'
import { createClient } from '~helpers'
import { mockedTransactionHistory } from '~domains/account/__mocks__/get-cash-transactions'
import { getCashTransactions } from '../get-cash-transactions'
import { loadFeatureFlag } from '~helpers/server-only/featureFlagClient'

describe(getCashTransactions, () => {
  describe('transactions - FF off', () => {
    beforeEach(() => {
      ;(axios.get as jest.Mock).mockResolvedValueOnce({
        data: mockedTransactionHistory
      })
      ;(loadFeatureFlag as jest.Mock).mockResolvedValue({})
    })

    const data = {
      locale: 'en-gb',
      params: { limit: '8', offset: '0' },
      accessToken: 'some-token'
    }

    it('creates the client and passes the locale and access token to ledger', async () => {
      await getCashTransactions(data)

      expect(createClient).toHaveBeenCalledWith({
        locale: 'en-gb',
        options: {
          baseURL: 'someLedgerBaseUrl'
        }
      })

      expect(axios.get).toHaveBeenCalledWith(
        `/ledger-data?limit=${data.params.limit}&offset=${data.params.offset}`,
        {
          headers: { Authorization: 'Bearer some-token' }
        }
      )
    })

    it('returns the correct response', async () => {
      const getCashTransactionsResponse = await getCashTransactions(data)
      expect(getCashTransactionsResponse).toMatchSnapshot()
    })
  })
  describe('transactions - FF on', () => {
    beforeEach(() => {
      ;(axios.get as jest.Mock).mockResolvedValueOnce({
        data: mockedTransactionHistory
      })
      ;(loadFeatureFlag as jest.Mock).mockResolvedValue(
        mockedTransactionHistory
      )
    })

    const data = {
      locale: 'en-gb',
      params: { limit: '8', offset: '0' },
      accessToken: 'some-token'
    }

    it('does not create the client or call the ledger service endpoint', async () => {
      await getCashTransactions(data)

      expect(createClient).toHaveBeenCalledTimes(0)

      expect(axios.get).toHaveBeenCalledTimes(0)
    })

    it('returns the correct response', async () => {
      const getCashTransactionsResponse = await getCashTransactions(data)
      expect(getCashTransactionsResponse).toMatchSnapshot()
    })
  })
})
