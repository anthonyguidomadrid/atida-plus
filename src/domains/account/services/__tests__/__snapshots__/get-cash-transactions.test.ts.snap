// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCashTransactions transactions - FF off returns the correct response 1`] = `
Object {
  "totalPages": 4,
  "transactions": Array [
    Object {
      "date": "2022-10-04",
      "transactions": Array [
        Object {
          "amount": 200,
          "transaction_date": "2022-10-04T14:20:22.029764445Z",
          "transaction_type": "debit",
        },
        Object {
          "amount": 280,
          "expiry_date": "2023-10-04T14:20:22.029764445Z",
          "transaction_date": "2022-10-04T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
    Object {
      "date": "2022-07-23",
      "transactions": Array [
        Object {
          "amount": 280,
          "expiry_date": "2023-07-23T14:20:22.029764445Z",
          "transaction_date": "2022-07-23T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
    Object {
      "date": "2022-06-23",
      "transactions": Array [
        Object {
          "amount": 200,
          "transaction_date": "2022-06-23T14:20:22.029764445Z",
          "transaction_type": "debit",
        },
        Object {
          "amount": 280,
          "expiry_date": "2023-06-23T14:20:22.029764445Z",
          "transaction_date": "2022-06-23T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
    Object {
      "date": "2022-05-23",
      "transactions": Array [
        Object {
          "amount": 500,
          "expiry_date": "2023-05-23T14:20:22.029764445Z",
          "transaction_date": "2022-05-23T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
    Object {
      "date": "2022-04-23",
      "transactions": Array [
        Object {
          "amount": 280,
          "expiry_date": "2023-04-23T14:20:22.029764445Z",
          "transaction_date": "2022-04-23T14:20:22.029764445Z",
          "transaction_type": "debit",
        },
        Object {
          "amount": 200,
          "expiry_date": "2023-04-23T14:20:22.029764445Z",
          "transaction_date": "2022-04-23T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
  ],
}
`;

exports[`getCashTransactions transactions - FF on returns the correct response 1`] = `
Object {
  "totalPages": 4,
  "transactions": Array [
    Object {
      "date": "2022-10-04",
      "transactions": Array [
        Object {
          "amount": 200,
          "transaction_date": "2022-10-04T14:20:22.029764445Z",
          "transaction_type": "debit",
        },
        Object {
          "amount": 280,
          "expiry_date": "2023-10-04T14:20:22.029764445Z",
          "transaction_date": "2022-10-04T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
    Object {
      "date": "2022-07-23",
      "transactions": Array [
        Object {
          "amount": 280,
          "expiry_date": "2023-07-23T14:20:22.029764445Z",
          "transaction_date": "2022-07-23T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
    Object {
      "date": "2022-06-23",
      "transactions": Array [
        Object {
          "amount": 200,
          "transaction_date": "2022-06-23T14:20:22.029764445Z",
          "transaction_type": "debit",
        },
        Object {
          "amount": 280,
          "expiry_date": "2023-06-23T14:20:22.029764445Z",
          "transaction_date": "2022-06-23T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
    Object {
      "date": "2022-05-23",
      "transactions": Array [
        Object {
          "amount": 500,
          "expiry_date": "2023-05-23T14:20:22.029764445Z",
          "transaction_date": "2022-05-23T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
    Object {
      "date": "2022-04-23",
      "transactions": Array [
        Object {
          "amount": 280,
          "expiry_date": "2023-04-23T14:20:22.029764445Z",
          "transaction_date": "2022-04-23T14:20:22.029764445Z",
          "transaction_type": "debit",
        },
        Object {
          "amount": 200,
          "expiry_date": "2023-04-23T14:20:22.029764445Z",
          "transaction_date": "2022-04-23T14:20:22.029764445Z",
          "transaction_type": "credit",
        },
      ],
    },
  ],
}
`;
